File: src/morphic/dupfinder/scanner.py
"""
Background scan job management for the dupfinder web UI.

Handles running duplicate-detection scans in background threads and
converting results into JSON-serializable formats.
"""

from __future__ import annotations

import logging
import os
import threading
import time
import uuid
from dataclasses import dataclass, field

from morphic.shared.constants import (
    DEFAULT_BATCH_SIZE,
    DEFAULT_HASH_SIZE,
    DEFAULT_IMAGE_THRESHOLD,
    DEFAULT_NUM_FRAMES,
    DEFAULT_NUM_WORKERS,
    DEFAULT_VIDEO_THRESHOLD,
)
from morphic.shared.utils import format_duration, format_file_size
from morphic.dupfinder.images import ImageDuplicateFinder, ImageInfo
from morphic.dupfinder.videos import VideoDuplicateFinder, VideoInfo

logger = logging.getLogger(__name__)


# ── Data Structures ────────────────────────────────────────────────────────


@dataclass
class ScanJob:
    """Represents a running or completed scan job."""

    id: str
    folder: str
    scan_type: str  # "images", "videos", "both"
    status: str = "pending"
    progress: float = 0.0
    message: str = ""
    error: str | None = None
    # Results
    image_groups: list[list[dict]] = field(default_factory=list)
    video_groups: list[list[dict]] = field(default_factory=list)
    image_infos: dict[str, ImageInfo] = field(default_factory=dict)
    video_infos: dict[str, VideoInfo] = field(default_factory=dict)
    total_files_found: int = 0
    total_files_processed: int = 0
    space_savings: int = 0
    started_at: float = 0.0
    finished_at: float = 0.0
    # Settings
    image_threshold: float = DEFAULT_IMAGE_THRESHOLD
    video_threshold: float = DEFAULT_VIDEO_THRESHOLD


# ── Job Registry ───────────────────────────────────────────────────────────

_scan_jobs: dict[str, ScanJob] = {}
_scan_lock = threading.Lock()


def get_job(job_id: str) -> ScanJob | None:
    """Retrieve a scan job by ID."""
    with _scan_lock:
        return _scan_jobs.get(job_id)


def start_job(
    folder: str,
    scan_type: str,
    image_threshold: float = DEFAULT_IMAGE_THRESHOLD,
    video_threshold: float = DEFAULT_VIDEO_THRESHOLD,
) -> str:
    """Create and launch a new scan job. Returns the job ID."""
    job_id = str(uuid.uuid4())[:8]
    job = ScanJob(
        id=job_id,
        folder=folder,
        scan_type=scan_type,
        image_threshold=image_threshold,
        video_threshold=video_threshold,
    )
    with _scan_lock:
        _scan_jobs[job_id] = job

    thread = threading.Thread(target=_run_scan, args=(job,), daemon=True)
    thread.start()
    return job_id


# ── Background Worker ──────────────────────────────────────────────────────


def _run_scan(job: ScanJob) -> None:
    """Execute the scan in a background thread."""
    try:
        job.status = "scanning"
        job.started_at = time.time()
        job.message = f"Scanning folder: {job.folder}"

        # ── Image Scan ─────────────────────────────────────────────
        if job.scan_type in ("images", "both"):
            job.message = "Finding image files..."
            finder = ImageDuplicateFinder(
                similarity_threshold=job.image_threshold,
                hash_size=DEFAULT_HASH_SIZE,
                num_workers=DEFAULT_NUM_WORKERS,
                use_gpu=True,
                batch_size=DEFAULT_BATCH_SIZE,
            )
            image_files = finder.find_images(job.folder)
            job.total_files_found += len(image_files)
            job.message = (
                f"Found {len(image_files)} images. Processing hashes..."
            )
            job.progress = 0.1

            if image_files:
                job.status = "processing"
                finder.process_images(image_files)
                job.image_infos = finder.image_infos
                job.total_files_processed += len(finder.image_infos)
                job.progress = (
                    0.4 if job.scan_type == "both" else 0.7
                )
                job.message = (
                    f"Processed {len(finder.image_infos)} images. "
                    "Finding duplicates..."
                )

                job.status = "comparing"
                groups = finder.find_duplicates()
                job.image_groups = _format_image_groups(
                    groups, finder.image_infos,
                )
                job.progress = (
                    0.5 if job.scan_type == "both" else 0.95
                )

        # ── Video Scan ─────────────────────────────────────────────
        if job.scan_type in ("videos", "both"):
            job.message = "Finding video files..."
            vfinder = VideoDuplicateFinder(
                similarity_threshold=job.video_threshold,
                num_frames=DEFAULT_NUM_FRAMES,
                hash_size=DEFAULT_HASH_SIZE,
                num_workers=DEFAULT_NUM_WORKERS,
                use_gpu=True,
            )
            video_files = vfinder.find_videos(job.folder)
            job.total_files_found += len(video_files)
            job.message = (
                f"Found {len(video_files)} videos. Processing hashes..."
            )
            job.progress = (
                0.6 if job.scan_type == "both" else 0.1
            )

            if video_files:
                job.status = "processing"
                vfinder.process_videos(video_files)
                job.video_infos = vfinder.video_infos
                job.total_files_processed += len(vfinder.video_infos)
                job.progress = (
                    0.8 if job.scan_type == "both" else 0.7
                )
                job.message = (
                    f"Processed {len(vfinder.video_infos)} videos. "
                    "Finding duplicates..."
                )

                job.status = "comparing"
                groups = vfinder.find_duplicates()
                job.video_groups = _format_video_groups(
                    groups, vfinder.video_infos,
                )
                job.progress = 0.95

        # ── Finalise ───────────────────────────────────────────────
        job.space_savings = _calculate_space_savings(job)
        job.status = "done"
        job.progress = 1.0
        job.finished_at = time.time()
        elapsed = job.finished_at - job.started_at
        total_groups = len(job.image_groups) + len(job.video_groups)
        job.message = (
            f"Done! Found {total_groups} duplicate group(s) "
            f"in {elapsed:.1f}s. "
            f"Potential savings: {format_file_size(job.space_savings)}"
        )

    except Exception as e:
        logger.exception("Scan failed")
        job.status = "error"
        job.error = str(e)
        job.message = f"Error: {e}"
        job.finished_at = time.time()


# ── Result Formatters ──────────────────────────────────────────────────────


def _format_image_groups(
    groups: list[list[tuple[str, float]]],
    infos: dict[str, ImageInfo],
) -> list[list[dict]]:
    """Convert image duplicate groups to JSON-serializable dicts."""
    result = []
    for group in groups:
        formatted = []
        sorted_group = sorted(
            group,
            key=lambda x: infos.get(
                x[0], ImageInfo(path=""),
            ).file_size,
            reverse=True,
        )
        for path, similarity in sorted_group:
            info = infos.get(path)
            if info:
                formatted.append({
                    "path": path,
                    "filename": os.path.basename(path),
                    "directory": os.path.dirname(path),
                    "width": info.width,
                    "height": info.height,
                    "resolution": f"{info.width}x{info.height}",
                    "format": info.format or "Unknown",
                    "file_size": info.file_size,
                    "file_size_formatted": format_file_size(
                        info.file_size,
                    ),
                    "similarity": round(similarity * 100, 1),
                    "type": "image",
                })
        if len(formatted) > 1:
            result.append(formatted)
    return result


def _format_video_groups(
    groups: list[list[tuple[str, float]]],
    infos: dict[str, VideoInfo],
) -> list[list[dict]]:
    """Convert video duplicate groups to JSON-serializable dicts."""
    result = []
    for group in groups:
        formatted = []
        sorted_group = sorted(
            group,
            key=lambda x: infos.get(
                x[0], VideoInfo(path=""),
            ).file_size,
            reverse=True,
        )
        for path, similarity in sorted_group:
            info = infos.get(path)
            if info:
                formatted.append({
                    "path": path,
                    "filename": os.path.basename(path),
                    "directory": os.path.dirname(path),
                    "width": info.width,
                    "height": info.height,
                    "resolution": f"{info.width}x{info.height}",
                    "duration": info.duration,
                    "duration_formatted": format_duration(
                        info.duration,
                    ),
                    "fps": round(info.fps, 1),
                    "file_size": info.file_size,
                    "file_size_formatted": format_file_size(
                        info.file_size,
                    ),
                    "similarity": round(similarity * 100, 1),
                    "type": "video",
                })
        if len(formatted) > 1:
            result.append(formatted)
    return result


def _calculate_space_savings(job: ScanJob) -> int:
    """Sum up file sizes for all but the largest file in each group."""
    total = 0
    for group in job.image_groups + job.video_groups:
        sizes = [item["file_size"] for item in group]
        if len(sizes) > 1:
            total += sum(sorted(sizes)[:-1])
    return total

File: internal/dupfinder/scanner.go
package dupfinder

import (
	"fmt"
	"path/filepath"
	"sort"
	"sync"
	"time"

	"github.com/exterex/morphic/internal/shared"
)

// ScanJob represents a running or completed dupfinder job.
type ScanJob struct {
	shared.Job
	mu sync.Mutex

	Folder         string                     `json:"folder"`
	ScanType       string                     `json:"scan_type"` // "images", "videos", "both"
	ImageThreshold float64                    `json:"image_threshold"`
	VideoThreshold float64                    `json:"video_threshold"`
	ImageGroups    [][]map[string]interface{} `json:"image_groups,omitempty"`
	VideoGroups    [][]map[string]interface{} `json:"video_groups,omitempty"`
	TotalFound     int                        `json:"total_files_found"`
	TotalProcessed int                        `json:"total_files_processed"`
	SpaceSavings   int64                      `json:"space_savings"`
}

var store = shared.NewJobStore[ScanJob]()

func init() {
	store.StartCleanup(30*time.Minute, func(j *ScanJob) time.Time {
		return j.DoneAt
	})
}

// StartJob creates and launches a new dupfinder job.
func StartJob(folder, scanType string, imageThreshold, videoThreshold float64) string {
	job := &ScanJob{
		Job:            shared.NewJob(),
		Folder:         folder,
		ScanType:       scanType,
		ImageThreshold: imageThreshold,
		VideoThreshold: videoThreshold,
	}
	job.Status = shared.JobStatusRunning
	store.Set(job.ID, job)
	go runScan(job)
	return job.ID
}

// GetJob retrieves a job by ID.
func GetJob(id string) (*ScanJob, bool) {
	return store.Get(id)
}

func runScan(job *ScanJob) {
	defer func() {
		if r := recover(); r != nil {
			job.mu.Lock()
			job.Status = shared.JobStatusFailed
			job.Error = fmt.Sprintf("%v", r)
			job.DoneAt = time.Now()
			job.mu.Unlock()
		}
	}()

	job.mu.Lock()
	job.Message = fmt.Sprintf("Scanning folder: %s", job.Folder)
	job.mu.Unlock()

	// Image scan
	if job.ScanType == "images" || job.ScanType == "both" {
		scanImages(job)
	}

	// Check for cancellation between phases
	select {
	case <-job.Ctx().Done():
		job.mu.Lock()
		job.Status = shared.JobStatusCancelled
		job.DoneAt = time.Now()
		job.Message = "Scan was interrupted"
		job.mu.Unlock()
		return
	default:
	}

	// Video scan
	if job.ScanType == "videos" || job.ScanType == "both" {
		scanVideos(job)
	}

	// Check for cancellation before finalising
	select {
	case <-job.Ctx().Done():
		job.mu.Lock()
		job.Status = shared.JobStatusCancelled
		job.DoneAt = time.Now()
		job.Message = "Scan was interrupted"
		job.mu.Unlock()
		return
	default:
	}

	// Finalise
	job.mu.Lock()
	job.SpaceSavings = calculateSpaceSavings(job)
	job.Status = shared.JobStatusDone
	job.Progress = 1.0
	job.DoneAt = time.Now()
	elapsed := job.DoneAt.Sub(job.StartedAt).Seconds()
	totalGroups := len(job.ImageGroups) + len(job.VideoGroups)
	job.Message = fmt.Sprintf("Done! Found %d duplicate group(s) in %.1fs. Potential savings: %s",
		totalGroups, elapsed, shared.FormatFileSize(job.SpaceSavings))
	job.mu.Unlock()
}

func scanImages(job *ScanJob) {
	job.mu.Lock()
	job.Message = "Finding image files..."
	job.mu.Unlock()

	files, err := shared.FindImageFiles(job.Folder)
	if err != nil {
		return
	}

	job.mu.Lock()
	job.TotalFound += len(files)
	job.Message = fmt.Sprintf("Found %d images. Processing hashes...", len(files))
	job.Progress = 0.1
	job.mu.Unlock()

	if len(files) == 0 {
		return
	}

	infos := ProcessImages(job.Ctx(), files, shared.DefaultNumWorkers)

	// Return early if cancelled during hash processing
	select {
	case <-job.Ctx().Done():
		return
	default:
	}

	job.mu.Lock()
	job.TotalProcessed += len(infos)
	job.Progress = 0.4
	job.Message = fmt.Sprintf("Processed %d images. Finding duplicates...", len(infos))
	job.mu.Unlock()

	groups := FindImageDuplicates(infos, job.ImageThreshold)
	formatted := formatImageGroups(groups, infos)

	job.mu.Lock()
	job.ImageGroups = formatted
	if job.ScanType == "both" {
		job.Progress = 0.5
	} else {
		job.Progress = 0.95
	}
	job.mu.Unlock()
}

func scanVideos(job *ScanJob) {
	job.mu.Lock()
	job.Message = "Finding video files..."
	job.mu.Unlock()

	files, err := shared.FindVideoFiles(job.Folder)
	if err != nil {
		return
	}

	job.mu.Lock()
	job.TotalFound += len(files)
	job.Message = fmt.Sprintf("Found %d videos. Processing hashes...", len(files))
	if job.ScanType == "both" {
		job.Progress = 0.6
	} else {
		job.Progress = 0.1
	}
	job.mu.Unlock()

	if len(files) == 0 {
		return
	}

	infos := ProcessVideos(job.Ctx(), files, shared.DefaultNumFrames, shared.DefaultNumWorkers)

	// Return early if cancelled during hash processing
	select {
	case <-job.Ctx().Done():
		return
	default:
	}

	job.mu.Lock()
	job.TotalProcessed += len(infos)
	if job.ScanType == "both" {
		job.Progress = 0.8
	} else {
		job.Progress = 0.7
	}
	job.Message = fmt.Sprintf("Processed %d videos. Finding duplicates...", len(infos))
	job.mu.Unlock()

	groups := FindVideoDuplicates(infos, job.VideoThreshold)
	formatted := formatVideoGroups(groups, infos)

	job.mu.Lock()
	job.VideoGroups = formatted
	job.Progress = 0.95
	job.mu.Unlock()
}

func formatImageGroups(groups [][]DuplicateEntry, infos map[string]*ImageInfo) [][]map[string]interface{} {
	var result [][]map[string]interface{}
	for _, group := range groups {
		// Sort by file size descending
		sort.Slice(group, func(i, j int) bool {
			ai := infos[group[i].Path]
			aj := infos[group[j].Path]
			if ai == nil || aj == nil {
				return false
			}
			return ai.FileSize > aj.FileSize
		})

		var formatted []map[string]interface{}
		for _, entry := range group {
			info := infos[entry.Path]
			if info == nil {
				continue
			}
			formatted = append(formatted, map[string]interface{}{
				"path":                entry.Path,
				"filename":            filepath.Base(entry.Path),
				"directory":           filepath.Dir(entry.Path),
				"width":               info.Width,
				"height":              info.Height,
				"resolution":          fmt.Sprintf("%dx%d", info.Width, info.Height),
				"format":              info.Format,
				"file_size":           info.FileSize,
				"file_size_formatted": shared.FormatFileSize(info.FileSize),
				"similarity":          float64(int(entry.Similarity*1000)) / 10,
				"type":                "image",
			})
		}
		if len(formatted) > 1 {
			result = append(result, formatted)
		}
	}
	return result
}

func formatVideoGroups(groups [][]DuplicateEntry, infos map[string]*VideoInfo) [][]map[string]interface{} {
	var result [][]map[string]interface{}
	for _, group := range groups {
		sort.Slice(group, func(i, j int) bool {
			ai := infos[group[i].Path]
			aj := infos[group[j].Path]
			if ai == nil || aj == nil {
				return false
			}
			return ai.FileSize > aj.FileSize
		})

		var formatted []map[string]interface{}
		for _, entry := range group {
			info := infos[entry.Path]
			if info == nil {
				continue
			}
			formatted = append(formatted, map[string]interface{}{
				"path":                entry.Path,
				"filename":            filepath.Base(entry.Path),
				"directory":           filepath.Dir(entry.Path),
				"width":               info.Width,
				"height":              info.Height,
				"resolution":          fmt.Sprintf("%dx%d", info.Width, info.Height),
				"duration":            info.Duration,
				"duration_formatted":  shared.FormatDuration(info.Duration),
				"fps":                 float64(int(info.FPS*10)) / 10,
				"file_size":           info.FileSize,
				"file_size_formatted": shared.FormatFileSize(info.FileSize),
				"similarity":          float64(int(entry.Similarity*1000)) / 10,
				"type":                "video",
			})
		}
		if len(formatted) > 1 {
			result = append(result, formatted)
		}
	}
	return result
}

func calculateSpaceSavings(job *ScanJob) int64 {
	var total int64
	allGroups := append(job.ImageGroups, job.VideoGroups...)
	for _, group := range allGroups {
		var sizes []int64
		for _, item := range group {
			if s, ok := item["file_size"].(int64); ok {
				sizes = append(sizes, s)
			}
		}
		if len(sizes) > 1 {
			sort.Slice(sizes, func(i, j int) bool { return sizes[i] < sizes[j] })
			for _, s := range sizes[:len(sizes)-1] {
				total += s
			}
		}
	}
	return total
}
